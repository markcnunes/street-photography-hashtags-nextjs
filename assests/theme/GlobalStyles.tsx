import React from 'react';
import { css, Global } from '@emotion/react';
import theme from './theme';
import { normalize } from 'polished';

function GlobalStyles() {
  return (
    <Global
      styles={css`
        @import url('https://fonts.googleapis.com/css2?family=Open+Sans:wght@400;700&family=Work+Sans:wght@600&display=swap');
        ${normalize()},
        *,
          *::before,
          *::after {
          box-sizing: inherit;
        }
        html,
        body {
          padding: 0;
          margin: 0;
          width: 100%;
          font-family: ${theme.fonts.body};
          font-weight: ${theme.fontWheights.body[1]};
          box-sizing: border-box;
        }
        #__next {
          display: flex;
          flex-wrap: wrap;
          background: ${theme.colors.gradient};
          min-height: 100vh;
        }
        a {
          cursor: pointer;
        }
        h1,
        h2,
        h3,
        h4,
        h5 {
          font-family: ${theme.fonts.heading};
          font-weight: ${theme.fontWheights.heading[1]};
          letter-spacing: 1px;
        }
        .ReactModalPortal {
          position: relative;
          z-index: 9999;
        }

        .ReactModal__Body--open {
          overflow: hidden;
          .Toastify__toast-container {
            z-index: 10000;
          }
        }

        .ReactModal__Overlay {
          display: flex;
          width: 100%;
          height: 100%;
          background: rgba(0, 0, 0, 0.85) !important;
          align-items: center;
          justify-content: center;
        }
      `}
    />
  );
}

export default GlobalStyles;
