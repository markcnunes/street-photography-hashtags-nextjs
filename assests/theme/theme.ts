const colors = {
  primary: '#615cf5',
  secondary: '#684e79',
  muted: '#536075',
  black: '#1c1c27',
  white: '#fff',
};

const theme = {
  fonts: {
    body: `'Open Sans', Helvetica, Arial, sans-serif`,
    heading: `'Work Sans', Helvetica, Arial, sans-serif`,
  },
  fontSizes: [12, 14, 16, 18, 24, 32, 48, 64, 72],
  fontWheights: {
    body: [0, 400, 700],
    heading: [0, 600],
  },
  colors: {
    primary: colors.primary,
    secondary: colors.secondary,
    muted: colors.muted,
    black: colors.black,
    white: colors.white,
    gradient: `linear-gradient(90deg, ${colors.black} 0%, ${colors.muted} 100%)`,
  },
  radii: [0, 2, 4, 8],
};

export default theme;
