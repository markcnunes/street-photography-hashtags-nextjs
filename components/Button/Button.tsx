import styled from '@emotion/styled';
import { variant } from 'styled-system';
import { lighten, rgba, saturate } from 'polished';

interface IButton {
  /**
   * If `true`, the button will be disabled.
   */
  disabled?: boolean;
  /**
   * If `true`, the button will be full width.
   */
  fullWidth?: boolean;
  /**
   * Takes of 'only', 'withText', It will adapt the button to have an icon.
   */
  icon?: 'only' | 'withText';
  /**
   * Takes one of 'primary', 'secondary', 'tertirary', "black", "white". corresponds to buttons colors
   */
  color?: 'primary' | 'secondary' | 'muted' | 'black' | 'white';
  /**
   * Takes one of 'small', 'medium', 'large'. corresponds to buttons sizes
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * Takes one of 'text', 'outlined', 'contained'. corresponds to buttons variations
   */
  variant?: 'text' | 'outlined' | 'contained';
  /**
   * Takes a function
   */
  onClick?: (event: any) => void;
  /**
   * Allows more props. corresponds to ...other
   */
  [propName: string]: any;
}

const StyledButton = styled('button')<IButton>(
  (props) => ({
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: props.fullWidth ? 'center' : 'auto',
    border: 0,
    outline: 0,
    cursor: 'pointer',
    width: props.fullWidth ? '100%' : 'auto',
    opacity: props.disabled ? 0.5 : 1,
    borderRadius: 5,
    textDecoration: 'none',
    transition: 'background 0.3s ease',
    'svg, span': {
      marginRight: props.icon === 'withText' ? 5 : undefined,
    },
    '&:disabled': {
      cursor: 'auto',
    },
  }),
  // color
  ({ theme, color = 'primary' }) =>
    variant({
      prop: 'color',
      variants: {
        primary: {
          color: theme.colors.white,
          background: theme.colors[color],
          '&:hover': {
            background: saturate(1, lighten(0.05, theme.colors[color])),
          },
        },
        secondary: {
          color: theme.colors.white,
          background: theme.colors[color],
          '&:hover': {
            background: saturate(1, lighten(0.05, theme.colors[color])),
          },
        },
        muted: {
          color: theme.colors.white,
          background: theme.colors[color],
          '&:hover': {
            background: lighten(0.2, theme.colors[color]),
          },
        },
        black: {
          color: theme.colors.white,
          background: theme.colors[color],
          '&:hover': {
            background: lighten(0.2, theme.colors[color]),
          },
        },
        white: {
          color: theme.colors.black,
          background: theme.colors[color],
          '&:hover': {
            background: lighten(0.2, theme.colors[color]),
          },
        },
      },
    }),
  // size
  ({ theme }) =>
    variant({
      prop: 'size',
      variants: {
        small: {
          fontSize: theme.fontSizes[1],
          padding: `8px 16px`,
          borderRadius: theme.radii[2],
        },
        medium: {
          fontSize: theme.fontSizes[2],
          padding: `8px 16px`,
          borderRadius: theme.radii[2],
        },
        large: {
          fontSize: theme.fontSizes[3],
          padding: `12px 24px`,
          borderRadius: theme.radii[3],
        },
      },
    }),
  // variant
  ({ theme, color = 'primary' }) =>
    variant({
      prop: 'variant',
      variants: {
        contained: {
          '&:disabled': {
            background: rgba(theme.colors.muted, 0.5),
            color: theme.colors.white,
          },
        },
        outlined: {
          border: '1px solid currentColor',
          background: 'transparent',
          color: theme.colors[color],
          '&:disabled': {
            color: rgba(theme.colors.muted, 0.5),
          },
          '&:hover': {
            background: rgba(theme.colors[color], 0.05),
            '&:disabled': {
              background: 'transparent',
              color: rgba(theme.colors.muted, 0.5),
            },
          },
        },
        text: {
          background: 'transparent',
          color: theme.colors[color],
          '&:disabled': {
            color: rgba(theme.colors.muted, 0.5),
          },
          '&:hover': {
            background: 'transparent',
            textDecoration: 'underline',
          },
        },
      },
    }),
);

const Button = ({
  onClick,
  disabled,
  children,
  color = 'primary',
  fullWidth = false,
  size = 'medium',
  variant = 'contained',
  ...other
}: IButton) => {
  return (
    <StyledButton
      color={color}
      onClick={onClick}
      disabled={disabled}
      fullWidth={fullWidth}
      size={size}
      variant={variant}
      {...other}
    >
      {children}
    </StyledButton>
  );
};

export default Button;
