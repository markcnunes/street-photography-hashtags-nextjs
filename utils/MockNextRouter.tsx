import { RouterContext } from 'next/dist/next-server/lib/router-context';
import { NextRouter } from 'next/router';
import { ReactNode } from 'react';

interface MockNextRouterProps extends Partial<NextRouter> {
  children?: ReactNode;
}

/**
 * Override the default NextRouter render with different attributes.
 * e.g. change the path or locale for tests.
 */

const MockNextRouter = ({ children, ...props }: MockNextRouterProps) => {
  const mockRouter: NextRouter = {
    route: '/',
    pathname: '',
    query: {},
    asPath: '/',
    basePath: '',
    locale: 'en',
    locales: [''],
    defaultLocale: 'en',
    // domainLocales?: DomainLocales,
    isLocaleDomain: true,
    push: jest.fn(() => Promise.resolve(true)),
    replace: jest.fn(() => Promise.resolve(true)),
    reload: jest.fn(() => Promise.resolve(true)),
    back: jest.fn(() => Promise.resolve(true)),
    prefetch: jest.fn(),
    beforePopState: jest.fn(() => Promise.resolve(true)),
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn(),
    },
    isFallback: false,
    isReady: true,
    isPreview: false,
  };

  return (
    <RouterContext.Provider value={{ ...mockRouter, ...props }}>
      {children}
    </RouterContext.Provider>
  );
};

export default MockNextRouter;
