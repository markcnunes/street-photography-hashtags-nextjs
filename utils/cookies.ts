// import ReactGA from 'react-ga';

import GA4React from 'ga-4-react';

export const cookieConsent = 'sphCookieconsent';

/*
 * Check if cookie consentment is true before initializing and rendering it
 */
export const getCookie = (key: string): string | undefined => {
  var b = document.cookie.match('(^|;)\\s*' + key + '\\s*=\\s*([^;]+)');
  return b ? b.pop() : '';
};

export const gaCode = 'G-W7QKGLQ9YV';
const initializeCookies = () => {
  if (!window.location.href.includes('localhost')) {
    const ga4react = new GA4React(gaCode);
    ga4react.initialize();
    // reload page to get the track running
    window.location.reload();
  }
};

export const renderCookies = (): any => {
  getCookie(cookieConsent) === 'true' && initializeCookies();
};

export default initializeCookies;
