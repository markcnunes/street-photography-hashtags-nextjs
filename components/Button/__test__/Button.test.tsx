import { rgba } from 'polished';
import React from 'react';
import { render } from 'utils/test-utils';
import theme from 'assests/theme';

import Button from '../Button';

describe('<Button />', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Button />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('can render a text', () => {
    const { asFragment } = render(<Button>lorem ipsum</Button>);
    expect(asFragment().textContent).toBe('lorem ipsum');
  });
  /**
   * Props: Colors
   */
  describe('renders color:', () => {
    it('primary - default', () => {
      const color = 'primary';
      const { getByRole } = render(<Button />);
      const button = getByRole('button');

      expect(button).toHaveAttribute('color', color); // styled-system adds this attribute for color
      expect(button).toHaveStyle({
        color: theme.colors.white,
        background: theme.colors[color],
      });
    });
    it('secondary', () => {
      const color = 'secondary';
      const { getByRole } = render(<Button color={color} />);
      const button = getByRole('button');

      expect(button).toHaveStyle({
        color: theme.colors.white,
        background: theme.colors[color],
      });
    });
    it('muted', () => {
      const color = 'muted';
      const { getByRole } = render(<Button color={color} />);
      const button = getByRole('button');

      expect(button).toHaveStyle({
        color: theme.colors.white,
        background: theme.colors[color],
      });
    });
    it('black', () => {
      const color = 'black';
      const { getByRole } = render(<Button color={color} />);
      const button = getByRole('button');

      expect(button).toHaveStyle({
        color: theme.colors.white,
        background: theme.colors[color],
      });
    });
    it('white', () => {
      const color = 'white';
      const { getByRole } = render(<Button color={color} />);
      const button = getByRole('button');

      expect(button).toHaveStyle({
        color: theme.colors.black,
        background: theme.colors[color],
      });
    });
  });
  /**
   * Props: Disabled
   */
  describe('renders disabled:', () => {
    it('false', () => {
      const { getByRole } = render(<Button />);
      const button = getByRole('button');

      expect(button).not.toHaveAttribute('disabled');
    });
    it('true', () => {
      const { getByRole } = render(<Button disabled />);
      const button = getByRole('button');

      expect(button).toHaveAttribute('disabled', '');
      expect(button).toHaveStyle({
        cursor: 'auto',
        color: theme.colors.white,
        background: rgba(theme.colors.muted, 0.5),
      });
    });
  });
  /**
   * Props: FullWidth
   */
  describe('renders fullWidth:', () => {
    it('false', () => {
      const { getByRole } = render(<Button />);
      const button = getByRole('button');

      expect(button).toHaveStyle({
        width: 'auto',
      });
    });
    it('true', () => {
      const { getByRole } = render(<Button fullWidth />);
      const button = getByRole('button');

      expect(button).toHaveStyle({
        width: '100%',
      });
    });
  });
  /**
   * Props: Layout
   */
  describe('renders layout:', () => {
    it('contained - default', () => {
      const color = 'primary';
      const { getByRole } = render(<Button />);
      const button = getByRole('button');
      // should render same layout of primary button
      expect(button).toHaveStyle({
        color: theme.colors.white,
        background: theme.colors[color],
      });
    });
    it('outlined', () => {
      const color = 'primary';
      const { getByRole } = render(<Button variant='outlined' />);
      const button = getByRole('button');

      expect(button).toHaveStyle({
        border: '1px solid currentColor',
        background: 'transparent',
        color: theme.colors[color],
      });
    });
  });
  /**
   * Props: Size
   */
  describe('renders size:', () => {
    it('small', () => {
      const { getByRole } = render(<Button size='small' />);
      const button = getByRole('button');

      expect(button).toHaveStyle({
        fontSize: theme.fontSizes[1],
        padding: `8px 16px`,
        borderRadius: theme.radii[2] + 'px',
      });
    });
    it('medium - default', () => {
      const { getByRole } = render(<Button />);
      const button = getByRole('button');

      expect(button).toHaveStyle({
        fontSize: theme.fontSizes[2],
        padding: `8px 16px`,
        borderRadius: theme.radii[2] + 'px',
      });
    });
    it('large', () => {
      const { getByRole } = render(<Button size='large' />);
      const button = getByRole('button');

      expect(button).toHaveStyle({
        fontSize: theme.fontSizes[3],
        padding: `12px 24px`,
        borderRadius: theme.radii[3] + 'px',
      });
    });
  });
});
