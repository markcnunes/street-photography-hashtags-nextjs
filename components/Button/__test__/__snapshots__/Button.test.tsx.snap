// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matches snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: auto;
  -ms-flex-pack: auto;
  -webkit-justify-content: auto;
  justify-content: auto;
  border: 0;
  outline: 0;
  cursor: pointer;
  width: auto;
  opacity: 1;
  border-radius: 5px;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background 0.3s ease;
  transition: background 0.3s ease;
  color: #fff;
  background: #615cf5;
  font-size: 16px;
  padding: 8px 16px;
  border-radius: 4px;
}

.emotion-0:disabled {
  cursor: auto;
}

.emotion-0:hover {
  background: #706bff;
}

.emotion-0:disabled {
  background: rgba(83,96,117,0.5);
  color: #fff;
}

<button
    class="emotion-0 emotion-1"
    color="primary"
  />
</DocumentFragment>
`;
